mifosng-android/src/main/java/com/mifos/mifosxdroid/core/MifosBaseActivity.java

```diff
@@ -131,13 +132,21 @@ public void hideKeyboard(View view) {
 
     @Override
     public void logout() {
-        if (PrefManager.getPassCodeStatus()) {
-            startActivity(new Intent(this, PassCodeActivity.class));
-        } else {
-            PrefManager.clearPrefs();
-            startActivity(new Intent(this, SplashScreenActivity.class));
-        }
-        finish();
+        AlertDialog.Builder builder = new AlertDialog.Builder(this);
+        builder.setMessage(R.string.logout_message)
+                .setPositiveButton(R.string.action_yes, new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int id) {
+                        PrefManager.clearToken();
```

Comment:
Remove all the sharedpreferences here instead of using `clearToken()` method.

When one logs out, the application should clear all the saved 'SharedPreference values'. `clearToken()` will only save a blank passcode value and all the remaining sharedpreferences will remain intact. You can save the `Instance Domain` to make it convenient for the other users to login but it is already handled in `LoginActivity`.

Reply:
@tarun0 I thought of clearing all the shared preferences at first but then I realized that if a user logs out and the logs in again his identity is lost and he will be a new user to the application making him to set the 4-digit passcode on subsequent logins.I will clear all the preferences for now and would start working on the problem of preventing the reset of password. I have already created an issue for that in mifos mobile as well.

Reply:
` if a user logs out and the logs in again his identity is lost and he will be a new user to the application making him to set the 4-digit passcode on subsequent logins` 
Isn't is exactly how it should be? Why do you think passcode should be retained after logout?

Reply:
@tarun0 Let's take an example of Paytm,The user creates a 4 digit PIN which is asked before any operation involving banking.The 4 digit passcode can be used instead of a longer PIN to ensure authenticity. Once the user logs-out the application(does not mean he wont use the app again,so why need to reset the password  ) the password is still saved since the user is expected to use the app again.That's my opinion.If you feel this is irrelavent i would amend the code.

Reply:
Your point is not irrelevant or wrong. However, you are mistaking the passcode used in the Mifos App with the PIN used in banking apps. 

 - While the PIN is an added security measure in the baking apps, the passcode in Mifos is a tad inclined towards convenience as compared to security. It provides a way for the user to log into the app with a 4 digit passcode saving him from entering username and password every time he opens the app. Our intention to add such a feature is that unauthorized user (who has physical access to the phone) shouldn't use be allowed to use the app when the authenticated user has already exited it (in the future versions, it's supposed to evolve into a feature that would ask for passcode every time before the user performs any writable action, e.g. apply for a loan).

 - The PIN used in banking apps is stored on the server side to verify the authenticity whereas, in our app, the Passcode is saved in SharedPreference (hashed form) only and the verification is done locally, without any kind of intervention from the server.

Still, there's some flaw in your proposition. If the user logs in from device 1 and saves the passcode and then logs out, assuming that the passcode still refers that particular user, what will happen if he logs in from device 2? How will device 1 get information that a new passcode has been generated on some other device? The solution to this intervention of the server.

Your proposition was something in the pipeline but it might affect offline functionality and would add to more network calls. As I'm not the right person to confirm or reject any possibility whether this feature will be added, I can only tell you that it's not present currently.

I hope it provides some clarification. :)

Reply:
@tarun0 Alright,I understood the functionality of the passcode now. Moreover retaining the passcode would be consistent only if the application is hosted on a universally available database on cloud. I have made the changes by clearing the preferences on logout.Thanks